src/main/java/com/google/devtools/build/lib/syntax/StringModule.java

```diff
@@ -368,58 +378,25 @@ public String replace(
   public MutableList<String> rsplit(
       String self, String sep, Object maxSplitO, Location loc, Environment env)
       throws EvalException {
-    int maxSplit = Type.INTEGER.convertOptional(maxSplitO, "'split' argument of 'split'", null, -1);
-    try {
-      return stringRSplit(self, sep, maxSplit, env);
-    } catch (IllegalArgumentException ex) {
-      throw new EvalException(loc, ex);
-    }
-  }
-
-  /**
-   * Splits the given string into a list of words, using {@code separator} as a delimiter.
-   *
-   * <p>At most {@code maxSplits} will be performed, going from right to left.
-   *
-   * @param input The input string.
-   * @param separator The separator string.
-   * @param maxSplits The maximum number of splits. Negative values mean unlimited splits.
-   * @return A list of words
-   * @throws IllegalArgumentException
-   */
-  private static MutableList<String> stringRSplit(
-      String input, String separator, int maxSplits, Environment env) {
-    if (separator.isEmpty()) {
-      throw new IllegalArgumentException("Empty separator");
+    if (sep.isEmpty()) {
+      throw new EvalException(loc, "Empty separator");
     }
-
-    if (maxSplits <= 0) {
-      maxSplits = Integer.MAX_VALUE;
+    int maxSplit = Integer.MAX_VALUE;
+    if (maxSplitO != Runtime.NONE) {
+      maxSplit = (Integer) maxSplitO;
     }
-
-    ArrayDeque<String> result = new ArrayDeque<>();
-    String[] parts = input.split(Pattern.quote(separator), -1);
-    int sepLen = separator.length();
-    int remainingLength = input.length();
-    int splitsSoFar = 0;
-
-    // Copies parts from the array into the final list, starting at the end (because
-    // it's rsplit), as long as fewer than maxSplits splits are performed. The
-    // last spot in the list is reserved for the remaining string, whose length
-    // has to be tracked throughout the loop.
-    for (int pos = parts.length - 1; (pos >= 0) && (splitsSoFar < maxSplits); --pos) {
-      String current = parts[pos];
-      result.addFirst(current);
-
-      ++splitsSoFar;
-      remainingLength -= sepLen + current.length();
-    }
-
-    if (splitsSoFar == maxSplits && remainingLength >= 0)   {
-      result.addFirst(input.substring(0, remainingLength));
+    ArrayDeque<String> res = new ArrayDeque<>();
+    int end = self.length() - 1;
+    while (true) {
+      int start = self.lastIndexOf(sep, end);
+      if (start < 0 || maxSplit-- == 0) {
+        res.addFirst(self.substring(0, Math.max(0, end + 1)));
+        break;
+      }
+      res.addFirst(self.substring(start + sep.length(), end + 1));
+      end = start - 1;
     }
-
-    return MutableList.copyOf(env, result);
+    return MutableList.copyOf(env, res);
```

Comment:
same here: use a MutableList, and simply reverse it before returning the result.

Reply:
But MutableList does not support `set`, so reversing it will still require a copy.
MutableList also does not support `add`, it supports `append` instead which needs `loc` and `env` arguments.

Reply:
Quite; I forgot that MutableList doesn't support any of the standard update operations and defines its own variants. wrapUnsafe looks like a good compromise.